### PR TITLE
Part1

### PR DESCRIPTION
**Pull request by dpsproservices**
Completed instruction steps 1 through 8. Game runs displays stats.

### CODE REVIEW

**Comment by dpsproservices** [comment]
Ok the game results match the expected output now.

File: Character.cpp [hunk 1]
```diff
@@ -97,21 +188,12 @@ void Character::attackInternal(Character& other)
             a) your stats are restored to their initial value if they are lower than it.
             b) your stats are boosted 10%
             c) the initial value of your stats is updated to reflect this boosted stat for the next time you defeat another character.
-      */
-        assert(false);
+        */
+
+        levelUpHitPoints();
+        levelUpArmor();
+        levelUpAttackDamage();
```
Comment (matkatmusic) [inline Character.cpp]: All of these functions do the SAME thing, just with different variable names.
use a SINGLE free function to do the same thing.  pass your params by reference.
- [ ] revise.

File: Character.cpp [hunk 2]
```diff
@@ -86,8 +125,60 @@ int Character::takeDamage(int damage)
     return hitPoints;
 }
 
+void Character::levelUpHitPoints()
+{
+    if(*initialHitPoints.get())
```
Comment (matkatmusic) [inline Character.cpp]: `get()` returns a pointer, which may be nullptr.  Then you're dereferencing it, which may be a nullptr dereference.
Then, because `initialHitPoints` owns an `int`, you're basically checking if the owned value is not zero.
All of this indicates you do not understand how to check if a raw pointer or `unique_ptr<>` instance is safe to use or not.

File: Character.h [hunk 3]
```diff
@@ -10,65 +10,73 @@
...
-    
-    const std::vector<std::unique_ptr<Item>>& getHelpfulItems() const { return helpfulItems; }
-    const std::vector<std::unique_ptr<Item>>& getDefensiveItems() const { return defensiveItems; }
+    int getHP() const;
+
+    int getArmorLevel() const;
+
+    int getAttackDamage() const;
+
+    bool getIsDefending() const;
```
Comment (matkatmusic) [inline Character.h]: You did not need to revise any of these.  Put them back the way they were.
- [ ] revise.

File: Character.h [hunk 4]
```diff
@@ -10,65 +10,73 @@
...
+
+    const std::vector<std::unique_ptr<Item>>& getHelpfulItems() const;
+
+    const std::vector<std::unique_ptr<Item>>& getDefensiveItems() const;
```
Comment (matkatmusic) [inline Character.h]: You did not need to add these functions.
- [ ] remove them.

File: Character.h [hunk 5]
```diff
@@ -10,65 +10,73 @@
...
-
-    void printStats()
-    {
-        std::cout << getName() << "'s stats: " << std::endl;
-        std::cout << getStats(); //make your getStats() use a function from the Utility.h
-        
-        std::cout << std::endl;
-        std::cout << std::endl;
-    }
+    void boostAttackDamage( int amount );
```
Comment (matkatmusic) [inline Character.h]: You did not need to move these to the cpp file.  put them back.
- [ ] revise.

File: Character.h [hunk 6]
```diff
@@ -10,65 +10,73 @@
...
     
     void attackInternal(Character& other);
-};
\ No newline at end of file
+
+    void levelUpHitPoints();
+
+    void levelUpArmor();
+    
+    void levelUpAttackDamage();
```
Comment (matkatmusic) [inline Character.h]: None of this needed editing.  Put it back the way it was.
- [ ] revise.

File: Dragon.cpp [hunk 7]
```diff
@@ -1,11 +1,19 @@
...
+#include "Character.h"
+#include "Utility.h"
 
-//Dragon::Dragon
+Dragon::Dragon
+(
+    std::string name_, 
+    int hitPoints_, 
+    int armor_
+) : Character(hitPoints_, armor_, 80), name(name_) {}
```
Comment (matkatmusic) [inline Dragon.cpp]: Please never write functions that looks like this.

The following is 100x easier to read.
```
Dragon::Dragon( std::string name_, int hitPoints_,  int armor_ ) : 
Character(hitPoints_, armor_, 80), 
name(name_) 
{

}
```

File: Dragon.h [hunk 8]
```diff
@@ -1,11 +1,24 @@
...
-private:
-    const std::string name;
-};
+{   
+    Dragon
+    (
+        std::string name_, 
+        int hitPoints_, 
+        int armor_
+    );
```
Comment (matkatmusic) [inline Dragon.h]: Do not put `(` on its own line.  you are confusing them with `{ }`
- [ ] revise

File: DragonSlayer.cpp [hunk 9]
```diff
@@ -1,28 +1,64 @@
...
+#include "Utility.h"
+#include "AttackItem.h"
 
-//DragonSlayer::DragonSlayer
+DragonSlayer::DragonSlayer
+(
+    std::string name_, 
+    int hitPoints_, 
+    int armor_
+) : Character(hitPoints_, armor_, 4), name(name_)
```
Comment (matkatmusic) [inline DragonSlayer.cpp]: - [ ] revise

File: DragonSlayer.h [hunk 10]
```diff
@@ -1,8 +1,29 @@
...
-private:
-    const std::string name;
-};
\ No newline at end of file
+    DragonSlayer
+    (
+        std::string name_, 
+        int hitPoints_, 
+        int armor
+    );
```
Comment (matkatmusic) [inline DragonSlayer.h]: - [ ] revise

File: DragonSlayer.h [hunk 11]
```diff
@@ -1,8 +1,29 @@
...
+    std::string getStats() override;
+
+    void attack( Character& other ) override;
+
+    private:
+
+        const std::string name;
+
+        // only DragonSlayer has an attack item the sword named Excalibur!
+        std::unique_ptr<AttackItem> attackItem { new AttackItem() };
```
Comment (matkatmusic) [inline DragonSlayer.h]: Fix this indentation
- [ ] revise

File: DragonSlayer.h [hunk 12]
```diff
@@ -1,8 +1,29 @@
 #pragma once
 
+#include <string>
+#include "Character.h"
```
Comment (matkatmusic) [inline DragonSlayer.h]: Check if Character includes `<string>` already.
- [ ] revise.

File: Dwarf.cpp [hunk 13]
```diff
@@ -1,7 +1,30 @@
...
+#include "Character.h"
+#include "Utility.h"
 
-//Dwarf::Dwarf
+Dwarf::Dwarf
+(
+    std::string name_, 
+    int hitPoints_, 
+    int armor_
+) : Character(hitPoints_, armor_, 4), name(name_)
```
Comment (matkatmusic) [inline Dwarf.cpp]: - [ ] REVISE

File: Dwarf.h [hunk 14]
```diff
@@ -1,7 +1,22 @@
...
+    );
+
+    const std::string& getName() override;
+
+    std::string getStats() override;
+
+    private:
+
+        const std::string name;
+};
```
Comment (matkatmusic) [inline Dwarf.h]: Revise.  See all of my previous comments about indentation, the way constructors should look, etc..
- [ ] REVISE

File: Item.cpp [hunk 15]
```diff
@@ -0,0 +1,7 @@
+#include "Item.h"
+
+Item::Item
+(
+    const std::string& name_, 
+    int effect 
+) : name(name_), boost(effect) { }
```
Comment (matkatmusic) [inline Item.cpp]: - [ ] REVISE

File: Item.h [hunk 16]
```diff
@@ -5,13 +5,19 @@
...
-    int boost;
-};
\ No newline at end of file
+
+    private:
+
+        std::string name;
+        
+        int boost;
+};
```
Comment (matkatmusic) [inline Item.h]: - [ ] Revise

File: Paladin.cpp [hunk 17]
```diff
@@ -1,7 +1,30 @@
...
+#include "Utility.h"
 
-// Paladin::Paladin
+Paladin::Paladin
+(
+    std::string name_, 
+    int hitPoints_, 
+    int armor_
+) : Character(hitPoints_, armor_, 10), name(name_) 
+{
```
Comment (matkatmusic) [inline Paladin.cpp]: - [ ] REVISE

File: Paladin.h [hunk 18]
```diff
@@ -1,7 +1,22 @@
...
+        int armor_
+    );
+
+    const std::string& getName() override;
+
+    std::string getStats() override;
+
+    private:
+
+        const std::string name;
```
Comment (matkatmusic) [inline Paladin.h]: - [ ] REVISE

File: Utility.cpp [hunk 19]
```diff
@@ -1,15 +1,20 @@
...
+#include "DragonSlayer.h"
+#include "Dragon.h"
+#include "Dwarf.h"
+#include "Paladin.h"
+#include "Item.h"
 #include "HelpfulItem.h"
 #include "DefensiveItem.h"
-#include "Character.h"
-
+#include "AttackItem.h"
```
Comment (matkatmusic) [inline Utility.cpp]: Some of these don't need to be included.
- [ ] REVISE

File: Utility.cpp [hunk 20]
```diff
@@ -31,6 +36,19 @@ std::vector<std::unique_ptr<Item>> makeDefensiveItems(int num)
...
+    std::vector<std::unique_ptr<Item>> items;
+    
+    while( --num >= 0 )
+    {
+        items.push_back( std::unique_ptr<AttackItem>(new AttackItem()) );
+    }
+    
+    std::cout << "made " << items.size() << " attack items" << std::endl;
+    return items;
+}
```
Comment (matkatmusic) [inline Utility.cpp]: Why did you add this function if you don't use it?
- [ ] remove it.

File: Utility.cpp [hunk 21]
```diff
@@ -1,15 +1,20 @@
...
-#include "Character.h"
-
+#include "AttackItem.h"
 
 std::vector<std::unique_ptr<Item>> makeHelpfulItems(int num)
 {
     std::vector<std::unique_ptr<Item>> items;
     
-    while( num-- >= 0 )
+    while( --num >= 0 )
```
Comment (matkatmusic) [inline Utility.cpp]: Why did you change this?
- [ ] revise.

File: Utility.cpp [hunk 22]
```diff
@@ -22,7 +27,7 @@ std::vector<std::unique_ptr<Item>> makeDefensiveItems(int num)
...
```
Comment (matkatmusic) [inline Utility.cpp]: Why did you change this?  
- [ ] revise

File: Utility.cpp [hunk 23]
```diff
@@ -43,66 +61,76 @@ std::string getCharacterStats(Character* ch)
...
         //this means you need to GET the attack damage, multiply it by the item's boost, and BOOST the attackDamage with that multiplied value.  
         //check Character.h for available member functions you can use.
+
+        int attackDamage = slayer->getAttackDamage(); // 4
+
+        int itemBoost = item->getBoost(); // 3 should be 10
+
+        int multiplier = attackDamage * itemBoost; // 4 x 3 = 12
+
+        slayer->boostAttackDamage(multiplier);
```
Comment (matkatmusic) [inline Utility.cpp]: Do this all with a single line of code.
- [ ] REVISE

File: Utility.h [hunk 24]
```diff
@@ -1,17 +1,19 @@
 #pragma once
 
+#include <string>
```
Comment (matkatmusic) [inline Utility.h]: Are you sure this needs to be included?
- [ ] REVISE

File: Utility.h [hunk 25]
```diff
@@ -1,17 +1,19 @@
...
-#include "Item.h"
-
 
+struct Item;
 std::vector<std::unique_ptr<Item>> makeHelpfulItems(int num);
 
 std::vector<std::unique_ptr<Item>> makeDefensiveItems(int num);
 
+std::vector<std::unique_ptr<Item>> makeAttackItems(int num);
+
```
Comment (matkatmusic) [inline Utility.h]: Instructions did not say to add this
- [ ] REVISE

File: Utility.h [hunk 26]
```diff
@@ -1,17 +1,19 @@
 #pragma once
 
+#include <string>
 #include <vector>
 #include <memory>
-#include "Item.h"
-
```
Comment (matkatmusic) [inline Utility.h]: Why did you remove this?
- [ ] REVISE

File: Utility.h [hunk 27]
```diff
@@ -1,17 +1,19 @@
...
 
+struct Item;
```
Comment (matkatmusic) [inline Utility.h]: Remove this forward declaration
- [ ] REVISE

**Review by matkatmusic — CHANGES_REQUESTED** [review]
You performed waaaaaaaay more code surgery than you needed to.   When you are given other people's code to work on, you need to spend more time thinking about if their existing code already works and if you need to revise it before you start diving in and ripping the guts out.

File: Utility.h (hunk 27 above)
Comment (dpsproservices) [reply Utility.h]: done

File: Utility.h (hunk 26 above)
Comment (dpsproservices) [reply Utility.h]: Figured only forward definition of struct Item is needed in Utility.h , because the definitions of the members have pointer and  reference and std::unique_ptr to type Item.

File: Character.h (hunk 3 above)
Comment (dpsproservices) [reply Character.h]: done

File: Character.h (hunk 5 above)
Comment (dpsproservices) [reply Character.h]: done

File: Utility.h (hunk 24 above)
Comment (dpsproservices) [reply Utility.h]: done it compiles and runs without it  that include

File: Character.h (hunk 4 above)
Comment (dpsproservices) [reply Character.h]: Moved implementation of getHelpfulItems() and getDefensiveItems() back to Character.h

File: Utility.h (hunk 25 above)
Comment (dpsproservices) [reply Utility.h]: done removed

File: Utility.cpp (hunk 23 above)
Comment (dpsproservices) [reply Utility.cpp]: done

File: Character.cpp (hunk 1 above)
Comment (dpsproservices) [reply Character.cpp]: combined into one levelUp()

File: Character.cpp (hunk 2 above)
Comment (dpsproservices) [reply Character.cpp]: ok

File: Character.h [hunk 28]
```diff
@@ -61,14 +68,30 @@ struct Character
...
     bool isDefending = false;
+
 private:
-    std::unique_ptr<int> initialHitPoints, initialArmorLevel, initialAttackDamage;
+
+    std::unique_ptr<int> initialHitPoints;
+
+    std::unique_ptr<int> initialArmorLevel;
+
+    std::unique_ptr<int> initialAttackDamage;
```
Comment (matkatmusic) [inline Character.h]: Why did you change all of this?

File: Character.h [hunk 29]
```diff
@@ -10,43 +10,50 @@
...
-    int getAttackDamage() const { return attackDamage; }
-    bool getIsDefending() const { return isDefending; }
-    
+    int getHP() const;
+
+    int getArmorLevel() const;
+
+    int getAttackDamage() const;
+
+    bool getIsDefending() const;
```
Comment (matkatmusic) [inline Character.h]: Why did you change this?

File: Item.cpp (hunk 15 above)
Comment (matkatmusic) [reply Item.cpp]: You still haven't fixed this ugly code style.
- [ ] REVISE

File: Item.h (hunk 16 above)
Comment (matkatmusic) [reply Item.h]: Still not fixed!

File: Utility.cpp (hunk 19 above)
Comment (matkatmusic) [reply Utility.cpp]: still not fixed!

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: Character.h (hunk 6 above)
Comment (dpsproservices) [reply Character.h]: done

File: Dragon.h (hunk 8 above)
Comment (dpsproservices) [reply Dragon.h]: ok done

File: Dragon.cpp (hunk 7 above)
Comment (dpsproservices) [reply Dragon.cpp]: ok done.

File: DragonSlayer.cpp (hunk 9 above)
Comment (dpsproservices) [reply DragonSlayer.cpp]: ok done

File: DragonSlayer.h (hunk 10 above)
Comment (dpsproservices) [reply DragonSlayer.h]: ok done.

File: DragonSlayer.h (hunk 12 above)
Comment (dpsproservices) [reply DragonSlayer.h]: ok done.

File: DragonSlayer.h (hunk 11 above)
Comment (dpsproservices) [reply DragonSlayer.h]: ok done.

File: Dwarf.cpp (hunk 13 above)
Comment (dpsproservices) [reply Dwarf.cpp]: done.

File: Item.cpp (hunk 15 above)
Comment (dpsproservices) [reply Item.cpp]: ok done.

File: Item.h (hunk 16 above)
Comment (dpsproservices) [reply Item.h]: done

File: Paladin.h (hunk 18 above)
Comment (dpsproservices) [reply Paladin.h]: done.

File: Utility.cpp (hunk 20 above)
Comment (dpsproservices) [reply Utility.cpp]: is removed.

File: Paladin.cpp (hunk 17 above)
Comment (dpsproservices) [reply Paladin.cpp]: done.

File: Utility.cpp (hunk 21 above)
Comment (dpsproservices) [reply Utility.cpp]: fixed number of helpful and defense items for each character.

File: Utility.cpp (hunk 22 above)
Comment (dpsproservices) [reply Utility.cpp]: fixed number of helpful and defense items for each character

File: Character.h (hunk 28 above)
Comment (dpsproservices) [reply Character.h]: Reverted.

File: Character.h (hunk 29 above)
Comment (dpsproservices) [reply Character.h]: Had moved member function implementations to cpp file instead of header file, is reverted back to the header file.

File: Utility.cpp (hunk 19 above)
Comment (dpsproservices) [reply Utility.cpp]: ok done.

**Review by matkatmusic — APPROVED** [review]

File: Dragon.cpp [hunk 30]
```diff
@@ -1,11 +1,19 @@
 #include "Dragon.h"
 #include "DragonSlayer.h"
+#include "Character.h"
```
Comment (matkatmusic) [inline Dragon.cpp:3]: Dragon.h is already including Character.h.

File: Item.cpp [hunk 31]
```diff
@@ -0,0 +1,7 @@
+#include "Item.h"
+
+Item::Item ( const std::string& name_, int effect ) : name(name_), boost(effect) { }
+
+Item::~Item() { }
+
+void Item::use(Character*) { }
```
Comment (matkatmusic) [inline Item.cpp]: Remove this default implementation.
- [ ] revise

File: Item.h [hunk 32]
```diff
@@ -5,13 +5,19 @@
...
 struct Item
 {
-    virtual ~Item() { }
-    virtual void use(Character*) = 0;
-    Item(const std::string& name_, int effect ) : name(name_), boost(effect) { }
+    Item(const std::string& name_, int effect );
     
+    virtual ~Item();
+
+    virtual void use(Character*);
```
Comment (matkatmusic) [inline Item.h]: This used to be pure virtual.  Why did you feel the need to change it?
- [ ] put it back.

File: Paladin.cpp [hunk 33]
```diff
@@ -1,7 +1,27 @@
 #include "Paladin.h"
+#include "Character.h"
```
Comment (matkatmusic) [inline Paladin.cpp]: Paladin.h is already including Character.h
- [ ] revise.

File: Dragon.cpp (hunk 30 above)
Comment (matkatmusic) [inline Dragon.cpp]: Dragon.h is already including Character.h
- [ ] revise

File: DragonSlayer.cpp [hunk 34]
```diff
@@ -1,28 +1,61 @@
 #include "DragonSlayer.h"
 #include "Dragon.h"
+#include "Character.h"
```
Comment (matkatmusic) [inline DragonSlayer.cpp]: DragonSlayer.h is already including Character.h

File: Dwarf.cpp [hunk 35]
```diff
@@ -1,7 +1,27 @@
 #include "Dwarf.h"
+#include "Character.h"
```
Comment (matkatmusic) [inline Dwarf.cpp]: Dwarf.h is already including Character.h
- [ ] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]
You seem to like to include things twice!
And why are there a bunch of extra new lines in classes you didn't make any other changes to?  Did you run this project through an auto-formatter or something when you finished it?
- [ ] answer in slack

File: Item.h (hunk 32 above)
Comment (dpsproservices) [reply Item.h]: done.

File: Dragon.cpp (hunk 30 above)
Comment (dpsproservices) [reply Dragon.cpp]: done.

File: DragonSlayer.cpp (hunk 34 above)
Comment (dpsproservices) [reply DragonSlayer.cpp]: done.

File: Dwarf.cpp (hunk 35 above)
Comment (dpsproservices) [reply Dwarf.cpp]: done.

File: Paladin.cpp (hunk 33 above)
Comment (dpsproservices) [reply Paladin.cpp]: done.

File: Item.cpp (hunk 31 above)
Comment (dpsproservices) [reply Item.cpp]: I had implemented the Item::use(Character*) in Item.cpp thinking it was causing vtables warnings.

The project compiles, links, runs without the implementation of void Item::use(Character*)  in Item.cpp

However commenting out the Item destructor in Item.cpp causes linker errors:

/tmp/Character-2d10d7.o: In function `Character::defend()':
Character.cpp:(.text+0x62a): undefined reference to `typeinfo for Item'
/tmp/Character-2d10d7.o: In function `Character::help(Character&)':
Character.cpp:(.text+0x7c1): undefined reference to `typeinfo for Item'
/tmp/AttackItem-25571b.o: In function `AttackItem::~AttackItem()':
AttackItem.cpp:(.text._ZN10AttackItemD2Ev[_ZN10AttackItemD2Ev]+0x11): undefined reference to `Item::~Item()'
/tmp/AttackItem-25571b.o:(.rodata+0x50): undefined reference to `typeinfo for Item'
/tmp/DefensiveItem-839137.o: In function `DefensiveItem::~DefensiveItem()':
DefensiveItem.cpp:(.text._ZN13DefensiveItemD2Ev[_ZN13DefensiveItemD2Ev]+0x11): undefined reference to `Item::~Item()'
/tmp/DefensiveItem-839137.o:(.rodata+0x50): undefined reference to `typeinfo for Item'
/tmp/HelpfulItem-51e61c.o: In function `HelpfulItem::~HelpfulItem()':
HelpfulItem.cpp:(.text._ZN11HelpfulItemD2Ev[_ZN11HelpfulItemD2Ev]+0x11): undefined reference to `Item::~Item()'
/tmp/HelpfulItem-51e61c.o:(.rodata+0x50): undefined reference to `typeinfo for Item'
/tmp/Item-d9f7b3.o: In function `Item::Item(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)':
Item.cpp:(.text+0xa): undefined reference to `vtable for Item'
clang-7: error: linker command failed with exit code 1 (use -v to see invocation)
exit status 1

**Comment by dpsproservices** [comment]
> And why are there a bunch of extra new lines in classes you didn't make any other changes to? Did you run this project through an auto-formatter or something when you finished it?

clang++ will output this warning below:

./Item.h:23:3: warning: C++98 requires newline at end of file [-Wc++98-compat-pedantic]

So I made sure every file has a newline on the last line in the repl.it editor.

See here:

https://stackoverflow.com/questions/72271/no-newline-at-end-of-file-compiler-warning

File: Item.cpp (hunk 31 above)
Comment (dpsproservices) [reply Item.cpp]: Compiles , links and runs fine now,  in Item.h the virtual destructor is defined but not implemented
the vtables warnings are gone.

Item.h
-------

virtual ~Item();  // semicolon terminated its defined but no curly braces { } means its not implemented here in Item.h


Item.cpp
---------

Item::~Item() { } // destructor implemented here curly braces { }

That's how I understand it/

**Review by matkatmusic — APPROVED** [review]
